docs/rholang/rholangtut.md

```diff
@@ -0,0 +1,501 @@
+     7     south!(spn)
+     8   } |
+     9   for (@spn <- south; @knf <- north) {
+    10     philosopher2!("Complete!") |
+    11     north!(knf) |
+    12     south!(spn)
+    13   }
+    14 }
+
+4, 9) The join operator, denoted with a semicolon `;`, declares that the continuation should only proceed if there is a message available on each of the channels simultaneously, preventing the deadlock above.
+
+## Crypto channels
+
+### Hashing
+
+There are three hashing functions available:
+- keccak256
+- sha256
+- blake2b256
+
+Hashing functions are exposed as channels which expect two arguments:
+- byte array to hash
+- return channel for sending back the hash
+
+
+#### Example usage:
+
+```rholang
+new x,y in {
+    x!(@"name"!("Joe") | @"age"!(40)) |  // (1)
+        for (@r <- x) {
+            @"keccak256Hash"!(r.toByteArray(), *y) // hash the program from (1)
+        } |
+        for (@h <- y) {
+            // the h here is hash of the rholang term we sent to the hash channel
+            // we can do anything we want with it but we choose to just print it
+            @"stdout"!(h)  // print out the keccak256 hash
+        }
+}
+```
+
+
+### Verify
+
+1. Let's hash some rholang program and print out it in base16. In rholang:
+```rholang
+new x,y in { 
+   x!(@"name"!("Joe") | @"age"!(40)) |  // (1)
+   for (@r <- x) { @"keccak256Hash"!(r.toByteArray(), *y) } |  // hash the program from (1)
+   for (@h <- y) { @"stdout"!(h) }  // print out the keccak256 hash
+}
+```
+
+This will print the hash of our program `(1)` : 
+`a6da46a1dc7ed715d4cd6472a736249a4d11142d160dbef9f20ae493de908c4e`
+
+2. We need pair of keys, let's generate some with `Ed25519` available in the project. In scala console:
+
+```scala
+import coop.rchain.crypto.signatures._
+import coop.rchain.crypto.codec._
+
+val keyPair = Ed25519.newKeyPair
+val secKey = Base16.encode(keyPair._1)
+// secKey: String = f6664a95992958bbfeb7e6f50bbca2aa7bfd015aec79820caf362a3c874e9247
+val pubKey = Base16.encode(keyPair._2)
+// pubKey: String = 288755c48c3951f89c5f0ffe885088dc0970fd935bc12adfdd81f81bb63d6219
+```
+
+3. Now we need to sign the hash we obtained in first step. Because we print out the base16 representation of underlying byte arrays for hashes we need to decode those strings.
+```
+val signature = Ed25519.sign(Base16.decode("a6da46a1dc7ed715d4cd6472a736249a4d11142d160dbef9f20ae493de908c4e"), Base16.decode(secKey))
+val base16Repr = Base16.encode(signature)
+// d0a909078ce8b8706a641b07a0d4fe2108064813ce42009f108f89c2a3f4864aa1a510d6dfccad3b62cd610db0bfe82bcecb08d813997fa7df14972f56017e0b
+```
+
+4. Now we can pass the signature and public key to our rholang program to verify it using crypto functions available. 
+
+`ed25519Verify` channel expects four arguments as follows:
+- data to verify. In our case this will be the keccak256 hash of our rholang program. The hash is represented in base16 so we need to call `hexToBytes` on it to turn the string into byte array
```

Comment:
In our case, this will be the keccak256 hash of our rholang program. The hash is represented in base16, so we need to call `hexToBytes` on it to turn the string into a byte array.

Reply:
Done.